homeassistant/components/ezviz/switch.py

```diff
@@ -24,61 +132,75 @@ async def async_setup_entry(
+    _attr_has_entity_name = True
 
     def __init__(
-        self, coordinator: EzvizDataUpdateCoordinator, serial: str, switch: str
+        self, coordinator: EzvizDataUpdateCoordinator, serial: str, switch_number: int
     ) -> None:
         """Initialize the switch."""
         super().__init__(coordinator, serial)
-        self._name = switch
-        self._attr_name = f"{self._camera_name} {DeviceSwitchType(switch).name.title()}"
+        self._switch_number = switch_number
         self._attr_unique_id = (
-            f"{serial}_{self._camera_name}.{DeviceSwitchType(switch).name}"
+            f"{serial}_{self._camera_name}.{DeviceSwitchType(switch_number).name}"
         )
-
-    @property
-    def is_on(self) -> bool:
-        """Return the state of the switch."""
-        return self.data["switches"][self._name]
+        self.entity_description = SWITCH_TYPES[switch_number]
+        self._attr_is_on = self.data["switches"][switch_number]
 
     async def async_turn_on(self, **kwargs: Any) -> None:
         """Change a device switch on the camera."""
         try:
             update_ok = await self.hass.async_add_executor_job(
-                self.coordinator.ezviz_client.switch_status, self._serial, self._name, 1
+                self.coordinator.ezviz_client.switch_status,
+                self._serial,
+                self._switch_number,
+                1,
             )
 
         except (HTTPError, PyEzvizError) as err:
-            raise PyEzvizError(f"Failed to turn on switch {self._name}") from err
+            raise HomeAssistantError(f"Failed to turn on switch {self.name}") from err
 
         if update_ok:
-            await self.coordinator.async_request_refresh()
+            self._attr_is_on = True
+            self.async_write_ha_state()
 
     async def async_turn_off(self, **kwargs: Any) -> None:
         """Change a device switch on the camera."""
         try:
             update_ok = await self.hass.async_add_executor_job(
-                self.coordinator.ezviz_client.switch_status, self._serial, self._name, 0
+                self.coordinator.ezviz_client.switch_status,
+                self._serial,
+                self._switch_number,
+                0,
             )
 
         except (HTTPError, PyEzvizError) as err:
-            raise PyEzvizError(f"Failed to turn off switch {self._name}") from err
+            raise HomeAssistantError(f"Failed to turn off switch {self.name}") from err
 
         if update_ok:
-            await self.coordinator.async_request_refresh()
+            self._attr_is_on = False
+            self.async_write_ha_state()
+
+    @callback
+    def _handle_coordinator_update(self) -> None:
+        """Handle updated data from the coordinator."""
+        if not self.data["switches"].get(self._switch_number):
+            return
+
+        self._attr_is_on = self.data["switches"][self._switch_number]
+        super()._handle_coordinator_update()
```

Comment:
Every 4 or 5 hours the API doesn't return switch statuses. (Happens about 2 or 3 times every 24 hours)

API also takes a few seconds to update the switch statuses from the actual device, so keeping track in memory until the next coordinator refresh seems to work best. This solves the switch flip/flop after set irritation.

Reply:
This would stop it from ever go `unavailable`.
Could we perhaps do a retry in the coordinator instead or something?